src/components/Loading.react.js

```diff
@@ -0,0 +1,165 @@
+                    status={loading_state}
+                    color={color}
+                    debug={debug}
+                    fullscreen={fullscreen}
+                />
+            );
+        }
+        if (
+            R.type(this.props.children) !== 'Object' ||
+            R.type(this.props.children) !== 'Function'
+        ) {
+            return <div className={className}>{this.props.children}</div>;
+        }
+        return this.props.children;
+    }
+}
+
+Loading.defaultProps = {
+    type: 'default',
+    color: '#119DFF',
+};
+
+Loading.propTypes = {
+    id: PropTypes.string,
+
+    /**
+     * Array that holds components to render
+     */
+    children: PropTypes.oneOfType([
+        PropTypes.arrayOf(PropTypes.node),
+        PropTypes.node,
+    ]),
+
+    /**
+     * Property that determines which spinner to show - one of 'graph', 'cube', 'circle', 'dot', or 'default'.
+     */
+    type: PropTypes.oneOf(['graph', 'cube', 'circle', 'dot', 'default']),
+
+    /**
+     * Boolean that determines if the loading spinner will be displayed full-screen or not
+     */
+    fullscreen: PropTypes.bool,
+
+    /**
+     * Boolean that determines if the loading spinner will display the status.prop_name and component_name
+     */
+    debug: PropTypes.bool,
+
+    /**
+     * Additional CSS class for the root DOM node
+     */
+    className: PropTypes.string,
+
+    /**
+     * Additional CSS styling for the root DOM node
+     */
+    style: PropTypes.object,
+
+    /**
+     * Primary colour used for the loading spinners
+     */
+    color: PropTypes.string,
+
+    /**
+     * Object that holds the loading state object coming from dash-renderer
+     */
+    loading_state: PropTypes.shape({
+        /**
+         * Determines if the component is loading or not
+         */
+        is_loading: PropTypes.bool,
+        /**
+         * Holds which property is loading
+         */
+        prop_name: PropTypes.string,
+        /**
+         * Holds the name of the component that is loading
+         */
+        component_name: PropTypes.string,
+    }),
+};
+
+function getLoadingStateInChildren(children, initial_loading_state) {
+    if (!Array.isArray(children)) {
+        children = [children];
+    }
+    let isLoading = initial_loading_state;
+    for (let i = 0; i < children.length; i++) {
+        const child = children[i];
+        // If we found another Loading component, we break,
+        // because the next Loading component will take care of it's
+        // own children
+        if (child.type === Loading) {
+            break;
+        }
+        if (
+            child.props &&
+            child.props.loading_state &&
+            child.props.loading_state.is_loading
+        ) {
+            isLoading = true;
+        }
+        if (child.props && child.props.children) {
+            isLoading = getLoadingStateInChildren(
+                child.props.children,
+                isLoading
```

Comment:
Passing the previous isLoading result seems to make the code more complicated for no gain. Would refactor it out.

Reply:
The `isLoading` here is needed because we don't want to lose any loading state when we're recursing through `child.props.children` - I'm having a hard time refactoring that out at least.